# Transfer learning in genome-wide association studies with knockoffs
This repository contains the code to reproduce the numerical results in Ren, Li, Sesia, and Sabatti (2021).

## Folders
* numerical_experiment_synthetic_data/: contains the code to reproduce the numerical results in Section 3.1 Numerical Experiments -- Synthetic data. To replicate the experiments, one needs to run the "run-me.sh" file on a computing cluster, and then run the "plot_results.R" file to make the plots. The "script.sh" file is specifically designed for the [Stanford Sherlock](https://www.sherlock.stanford.edu/) cluster. One may need to modify the files correspondingly. 
* numerical_experiment_gwas/: contains the code to reproduce the numerical results in Section 3.2 Numerical Experiments -- Real genotype data and simulated phenotypes. To replicate the experiments, one needs to have access to the [UK Biobank](https://www.ukbiobank.ac.uk/) genotype data as well as to the corresponding knockoffs generated by [Sesia et al. 2021](https://www.biorxiv.org/content/10.1101/2020.08.04.236703v2) ("FDR control in GWAS with population structure"). The script "choose_causal.R" is used to select the subsets of variants on chromosome one that will carry the causal signals for these experiments. The script "analysis.sh" fits the lasso model and computes the test statistics for each population. The script "analysis_transfer.sh" applies the weighted lasso transfer method, while the script "analysis_others.sh" applies the other methods. This code is specifically designed for the [Stanford Sherlock](https://www.sherlock.stanford.edu/) cluster. One may need to modify the files correspondingly. 
* .../:

## References
Zhimei Ren, Shuangning Li, Matteo Sesia, and Chiara Sabatti. <b>Transfer learning in genome-wide association studies with knockoffs</b>. 2021.
